ouroboros-consensus/test-consensus/Test/Dynamic/WIP-1202-PLAN.hs

```diff
@@ -20,12 +20,21 @@ Assumed by the test infrastructure:
 Guaranteed by the test infrastructure:
 
   * The test infrastructure maintains the *actual* delegation map.
+    
+    EdV: This is the only part of this plan that I feel needs some clarification. 
+    What does the "actual" delegation map mean, in the context of a blockchain with 
+    potential forks? Actual according to who? How does this relate to the
+    delegation map in the chain?
```

Comment:
I had the same question. I think the following is meant: the delegation map the honest core node operators want to be in place. The delegation map in the chain will lag behind. So this indeed needs some clarification.

Reply:
So are we going to assume the 7 core node operators are going to off-chain coordinate these changes? It's not an unreasonable assumption, perhaps; though perhaps it is, I'm not entirely sure. But if they don't coordinate, two core node operators may both independently delegate, at which point the "active" map  would be the map with both changes applied. But how "active" is already a lot murkier.

Reply:
> I think the following is meant: the delegation map the honest core node operators want to be in place. The delegation map in the chain will lag behind. 

Precisely right: It's the _actual_ map in the sense that it characterize the _actual_ delegations desired by each genesis key owner (i.e. some human being). And he's also right that the chains will lag behind that until they include the most recent delegation certificates.

I'll try to make that clearer.

> So are we going to assume the 7 core node operators are going to off-chain coordinate these changes?

Do multiple operators have to coordinate in order to decide to whom a particular genesis key has actually been delegated? I'd assumed that at most one node operator "owns" each genesis key.

Maybe "desired" instead of "actual" would be clearer?

> But if they don't coordinate, two core node operators may both independently delegate, at which point the "active" map would be the map with both changes applied. But how "active" is already a lot murkier.

I'm assuming you meant _actual_ not _active_. Related to my previous question: two node operators cannot "independently delegate" the same genesis key, can they? If not, then I think the notion of the "actual"/"desired" map seems easy to maintain.

Reply:
No, they would only make decisions on their own genesis keys, but nonetheless, that means that two operators can independently make changes to the delegation map, which means that the concept of a single global "active" delegation map may not exist, right?

Reply:
It seems that we're considering two different things. I haven't made any claims about what I think you mean by the "active" delegation map. I agree that there can be different delegation maps on different chains in the net.

I'm so far only concerned with an off-chain record of the latest intentions of all genesis key owners -- I called that the "actual" delegation map because it contains the latest delegations. It's clear now that the word "actual" is confusing. So I'm going to start writing "intended" delegation map instead. Does that make sense?

-----

Basic idea: The test infrastructure maintains an _intended delegation map_ that evolves over time, and it promptly issues delegation certificates when that map changes so that the nets' chains' ledgers' delegation maps will eventually be updated to equal the latest intended delegation map.

If we further assume that the test infrastructure immediately shuts down a node before it removes that node's relevant mapping in the intended delegation map, then we will ensure the invariant that we had previously decided was desirable:

```
  * Invariant: For all genesis keys @gk@, no two node instances simultaneously
    have a current ledger that maps @gk@ to their own operational key.
```

Reply:
I still find that confusing. I guess the confusion comes from the fact that there _isn't_ a _single_ intended/actual delegation map. Each node might have its own, no? If two nodes each make changes to the delegation map (which will become active at some point in the future), what exactly is _the_ intended delegation map?